[MRG+1] EHN/DOC Make error msg more informative in Pipeline

glemaitre opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->

None

#### What does this implement/fix? Explain your changes.

The message in `Pipeline` can be tricky to understand. If a `joblib.Memory` is passed instead of
`sklearn.externals.joblib.Memory`, the error message mention that we should provide a `joblib.Memory`, we got `memory=Memory(...)` instead.

This PR intends to:

- Make the error more explicit by printing the type of the object passed in the error message;
- Write explicitly in the documentation that `Memory` has to be from `sklearn.externals.joblib`

#### Any other comments?

Another solution is to make some ducktyping: we can remove the `isinstance` by checking that the object has a known `joblib.Memory` method. However, it might open the door to some bugs due to different versions of internal and external joblib.



<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

glemaitre committed:
EHN/DOC Make error msg more informative in Pipeline

glemaitre committed:
FIX python 2.7 error

glemaitre commented:
The [artifact](https://10213-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/generated/sklearn.pipeline.Pipeline.html#sklearn.pipeline.Pipeline) showing that the numpydoc is fine.

@jnothman @lesteve any thoughts?

lesteve commented:
I seem to remember there were other instances of `joblib.Memory` in docstrings and/or error messages in other places in the code. Maybe we should fix them in the same PR ?

An alternative is to check for a `cache` attribute that is a method. This is probably good enough for caching purposes.

Not sure whether one is strongly preferable over the other.

glemaitre commented:
> Maybe we should fix them in the same PR ?

Agreed, I'll check that.

glemaitre committed:
DOC/FIX fix all occurences of memory

glemaitre commented:
I fixed the comment of @lesteve. Also, I `memory` has a mutable argument in `randomized_l1` which I corrected using `None` as default like for the other cases.

glemaitre committed:
TST test caching in randomized_l1

jnothman reviewed (approved):
LGTM if you're finished.

glemaitre committed:
TST test memory error in randomized_l1

glemaitre commented:
@jnothman @TomDLT Done and all green.
